Render results based on devolved administration availability from the content item

Rosa-Fox opened the pull request:
⚠️ This repo is Continuously Deployed: make sure you [follow the guidance](https://docs.publishing.service.gov.uk/manual/development-pipeline.html#merge-your-own-pull-request) ⚠️

[Trello](https://trello.com/c/ahXDKAYp/402-update-frontend-for-local-transactions-with-devolved-administration-data).

# What?

This task is part of an epic to give publishers control over specifying devolved administration availability of a service. As a result of this work, we will render local transaction results pages based on devolved administration availability data from the content item. 

Currently the functionality to display different results pages based on devolved administration availability exists, but the availability data is read from the `local_transactions_config.yml` file. The existing functionality has been removed as part of this PR.

As well as now reading the devolved administration availability data from the content item, the views and content have been updated. See the [Trello](https://trello.com/c/ahXDKAYp/402-update-frontend-for-local-transactions-with-devolved-administration-data) card to understand what the design/content should be.

# Why?
At the moment, adding service data to the yml file involves developer intervention. Adding this as a feature to the publishing application will make things easier for content designers and developers. Although there are only 2 services specified in the yml file, we have identified that there are many other services that are unavailable to devolved administrations. Publishers will also be able to update these services via the publishing tool, which should make for better results to users.

### Screen shots 

#### Devolved administration availability with alternative URL (`devolved_administration_service.html.erb`)

![devolved-administration-service](https://user-images.githubusercontent.com/5963488/122916340-ae4e1b00-d354-11eb-97ce-058958d8b83a.png)

#### Service unavailable to devolved administration (`unavailable_service.html.erb`)

![unavailable-service](https://user-images.githubusercontent.com/5963488/122916385-bad27380-d354-11eb-8d68-3b9b27594305.png)


#### DA default behaviour - no alternative url provided (`results.html.erb`)

![da-no-alternative-url](https://user-images.githubusercontent.com/5963488/122916593-f66d3d80-d354-11eb-8c22-6f6b53fcb26f.png)



## Non devolved administration availability (`results.html.erb`)

![england-results](https://user-images.githubusercontent.com/5963488/122916430-c756cc00-d354-11eb-8325-a99910541327.png)

kevindew commented on app/controllers/local_transaction_controller.rb:
```diff
@@ -34,19 +34,27 @@ def results
+    @devolved_administration_availability = devolved_administration_availability
     @service = LocalTransactionService.new(
-      @publication.title, lgsl, @country_name, @local_authority.url
+      @publication.title, lgsl, @country_name, @local_authority.url, @devolved_administration_availability
```
Your reasoning about not wanting lots of arguments sounds good. I think rather than having a dedicated class for devolved_administration_availability it'd probably make sense to use the [LocalTransactionPresenter](https://github.com/alphagov/frontend/blob/master/app/presenters/local_transaction_presenter.rb) as a place to ask our devolved_administration questions. This is already set in this class and is @publication. We could then adapt the LocalTransactionService to accept the presenter as an argument rather than just the title - however I do question the need for LocalTransactionService in a later comment.

kevindew commented on app/models/local_transaction_service.rb:
```diff
@@ -1,41 +1,43 @@
-  def body
-    @unavailable_content["body"]
+  def unavailable?
+    find_devolved_administration_availability[country_name] == "unavailable"
```
If we had the presenter class available here I wonder if we could pass some of the work of this class that way.

For example this code could be:
```
def unavailable?
  @presenter.unavailable?(country_name)
end
```

and then in the presenter we have something like
```
def unavailabe?(country_name)
  key = country_name.downcase.gsub(" ", "_")
  @content_item.dig("#{key}_availability", "type") == "unavailable"
end
```
or similar - since we probably don't want this class to know too much about the content item details.

kevindew replied:
As I think about this more I wonder whether we're going to need this LocalTransactionService class by moving the devolved_administration stuff into the Presenter. I think we're going to leave this class in a strange place by adapting it as most of it's concerns seem to be based around unavailable handling - with much of it becoming redundant by this work.

I'm imagining we can probably adapt the controller so it's like:

```
    if @publication.unavailable?(@country_name)
      render :unavailable_service
    elsif @publication.devolved_administration?(@country_name)
      render :devolved_administration_service
    else
      render :results
    end
```

And enable us to delete this class

kevindew commented on app/models/local_transaction_service.rb:
```diff
@@ -1,41 +1,43 @@
-  def button_text
-    @unavailable_content["button_text"] || "Find other services"
+  def devolved_administration?
+    find_devolved_administration_availability[country_name] == "devolved_administration_service"
```
Similar point here that we could moved some of the logic here into the presenter.

kevindew commented on app/models/local_transaction_service.rb:
```diff
@@ -1,41 +1,43 @@
+    if devolved_administration?
+      alternative_url[country_name]
+    end
   end
```
I feel like this button_link method doesn't really belong in this class (it seems an odd choice historically) as it's very much a presentation concern. I'd suggest we get rid of it and have a `devolved_administration_alternative_url(country_name)` method on the presenter and call that directly from the devolved_administration view

kevindew commented on app/views/local_transaction/unavailable_service.html.erb:
```diff
@@ -8,27 +8,24 @@
-        <% end %>
-      </div>
-    </section>
-  <% end %>
```
This seems good to remove - it's seems optimistic that this will always complement the unavailable copy, and it will have just been on the previous page.

Rosa-Fox replied:
I have added this to the unavailable page. Do you think we would need it for the devolved administration page? I thought not at first as I am worried that England related content will display there. However there is additional info in the current YML file. It feels safer to leave it out...

kevindew replied:
Hmm yeah it's tricky to work out what's best. I don't think it should be on unavailable but not on devolved_administration_service - as I imagine any arguments that it should be on unavailable also apply to devolved_administration_service. 

I'd say either have it on both unavailable and devolved_administration_service or neither, I don't mind too much which both have pro's and con's so go with which one you're most comfortable with.

Rosa-Fox replied:
I removed them both for now. If design/content people decide that they want it, then it will be easy enough to add it in again.

kevindew replied:
Great stuff 👍 I think the last thing to do is to document that decision with the commit message.

Rosa-Fox replied:
👍

kevindew commented on config/environments/test.rb:
```diff
@@ -35,8 +35,4 @@
-
-  config.after_initialize do
-    Frontend.local_transactions_config = "test/fixtures/unavailable_services.yml"
-  end
```
Great stuff 👍

kevindew reviewed (commented):
Thanks Rosa - great to see this working.

I've focused most of my comments on the initial modelling where we've got 3 classes in play LocalTransactionPresenter, DevolvedAdministrationAvailability and LocalTransactionService. I'm wondering whether we can utilise the presenter class more and then not need either of the other classes. Hope that makes sense.

On the analytics front I wonder if we're going to want any event logging when the user views the page - rather than interacting with the button (so that we log they've got an unavailable.

Rosa-Fox commented:
@kevindew thanks for your review, it is ready for re-review. 

I have re-written it to make use of the LocalTransactionPresenter and it a way better approach.

I think it could also be good to get a review from a content writer/designer. I will take some screen grabs on Monday using integration and ask for them to look them over. I am not 100% confident due to the content example (word doc) not including as much text as the web pages we used for design reference. I am not sure if the lines not included in the content don't need changing, or if they shouldn't be included.

kevindew commented on test/functional/local_transaction_controller_test.rb:
```diff
@@ -137,6 +140,52 @@ class LocalTransactionControllerTest < ActionController::TestCase
+          post :search, params: { slug: "send-a-bear-to-your-local-council", postcode: "EH8 8DX" }
+        end
+
+        should "sanitize postcodes and redirect to the slug for the appropriate authority tier" do
```
I don't think this should is quite right for the scenario - particularly as the postcode isn't malformed. We should rephrase this more focused on the scenario we're testing.

Rosa-Fox replied:
I changed it to `should "redirect to the local authority slug"`

kevindew commented on test/integration/local_transactions_test.rb:
```diff
@@ -51,6 +51,8 @@ class LocalTransactionsTest < ActionDispatch::IntegrationTest
         lgil_override: 8,
```
Hmm there seems to be quite a high degree of overlap between the purpose of this integration test and the controller tests. It seems unfortunate but I'm not sure there's a quick way out of this problem - I guess this is a sign that there's either too many integration tests or that the controller tests are testing too much.

If we're not testing any scenarios in the controller tests that aren't covered by the integration tests we should just not add them.

Rosa-Fox replied:
Yeah I wasn't 100% sure. I did find it quite useful to have all of these integration tests, but I understand they are quite cumbersome. Is it ok to keep them?

kevindew replied:
Yeah they can be kept. There's probably no need for the country based functional tests as they're not really testing that isn't tested.

Rosa-Fox replied:
I removed the functional country based tests (I left a NI test in there,  but it was already there before I added anything so I decided to keep it there).

kevindew commented on test/integration/local_transactions_test.rb:
```diff
@@ -365,7 +367,7 @@ class LocalTransactionsTest < ActionDispatch::IntegrationTest
 
     should "show advisory message that we have no url" do
-      assert page.has_content?("We don't have a link for their website.")
+      assert page.has_content?("We don't have a link for their website")
```
Do we need to drop this period? I wasn't sure if it was actually a mistake.

Rosa-Fox replied:
I am not 100% sure but the [current results page](https://www.gov.uk/report-stray-dog/milton-keynes) doesn't have one, so I took it out.

kevindew replied:
I can't see that text on the example you linked (which means I'm pleased Milton Keynes does have a service for reporting stray dogs)

I can see the period in the source code though: https://github.com/alphagov/frontend/blob/a61c3bd90c60e11d50228b6b6983ed8b82746a00/app/views/local_transaction/results.html.erb#L49 unless we're somehow talking about different things?

kevindew commented on test/integration/local_transactions_test.rb:
```diff
@@ -403,91 +405,180 @@ class LocalTransactionsTest < ActionDispatch::IntegrationTest
+    end
+  end
+
+  context "when a service is available for a devolved administration" do
```
```suggestion
  context "when a service is handled differently for a devolved administration" do
```

I think we have to be careful we don't make this sound like the normal situation but for a devolved administration postcode.

Rosa-Fox replied:
👍

kevindew commented on test/integration/local_transactions_test.rb:
```diff
@@ -403,91 +405,180 @@ class LocalTransactionsTest < ActionDispatch::IntegrationTest
+      click_on "Find"
+    end
+
+    should "render devolved_administration_service view" do
```
We should avoid referencing the view name here since it is a user facing concern and instead focus this on the visible behaviour.

Rosa-Fox replied:
👍

kevindew commented on test/integration/local_transactions_test.rb:
```diff
@@ -403,91 +405,180 @@ class LocalTransactionsTest < ActionDispatch::IntegrationTest
+    end
+  end
+
+  context "when availability data is not provided for a devolved administration" do
```
I don't think we need this test as it doesn't exercise any different code paths. I think we should just make sure one of the earlier integration tests of the feature has a non England country name so we can implicitly cover it.

Rosa-Fox replied:
I removed it. I guess having the other tests for devolved administrations / the Westminster postcode for England for the tests at the top of the file cover this.

kevindew commented on test/unit/presenters/local_transaction_presenter_test.rb:
```diff
@@ -16,4 +24,37 @@ def subject(content_item)
   end
+
+  test "#scotland_availability" do
+    assert_equal "devolved_administration_service", subject(@content_item_with_devolved_administration_availability, "Scotland").scotland_availability["type"]
+    assert_equal "https://www.gov.scot/stray-dog", subject(@content_item_with_devolved_administration_availability, "Scotland").scotland_availability["alternative_url"]
```
You can probably make this test self-contained like the ones that precede it - which makes it easier to understand.

```suggestion
    devolved_administration = { "type" => "devolved_administration_service", "alternative_url" => "https://gov.scot" }
    assert_equal devolved_administration, subject(details: { scotland_availability: devolved_administration })
```

Rosa-Fox replied:
Ooh I did that as I wanted to make use of the content-schema example... but I suppose defining the structure in these tests saves someone having to go and find the schema. Have updated!

kevindew commented on test/unit/presenters/local_transaction_presenter_test.rb:
```diff
@@ -16,4 +24,37 @@ def subject(content_item)
+    assert_equal "https://www.nidirect.gov.uk/stray-dog", subject(@content_item_with_devolved_administration_availability, "Northern Ireland").northern_ireland_availability["alternative_url"]
+  end
+
+  test "#unavailable?" do
```
I think this should be broken up into tests for the different scenarios:

```
test "#unvailable? returns true for a devolved administration marked as unavailable"
test "#unvailable? returns false for a devolved administration marked as devolved_administration_service"
test "#unvailable? returns false for a non devolved administration"
```

I think these can also be self contained which should make them a little easier to understand.

Rosa-Fox replied:
👍

Rosa-Fox replied:
👍

kevindew commented on test/unit/presenters/local_transaction_presenter_test.rb:
```diff
@@ -16,4 +24,37 @@ def subject(content_item)
+    assert_equal false, subject(@content_item_with_devolved_administration_availability, "Northern Ireland").unavailable?
+  end
+
+  test "#devolved_administration?" do
```
Similar point about `#unavailable?` for being broken up.

Rosa-Fox replied:
👍

kevindew commented on test/unit/presenters/local_transaction_presenter_test.rb:
```diff
@@ -16,4 +24,37 @@ def subject(content_item)
+    assert_equal false, subject(@content_item_with_devolved_administration_availability, "Wales").devolved_administration?
+  end
+
+  test "#devolved_administration_alternative_url" do
```
Similar point about `#unavailable?` for being broken up.

kevindew reviewed (commented):
Thanks @Rosa-Fox this looks a lot cleaner without those extra classes. I've added a bunch of comments to refine this.

Should we look at handing this work over to another team member? I'm assuming it's unlikely you'll have a chance to look at this today.

Rosa-Fox commented on app/presenters/local_transaction_presenter.rb:
```diff
@@ -1,13 +1,53 @@
 class LocalTransactionPresenter < ContentItemPresenter
+  attr_reader :scotland_availability, :wales_availability, :northern_ireland_availability, :country_name
```
👍

Rosa-Fox commented on app/presenters/local_transaction_presenter.rb:
```diff
@@ -1,13 +1,53 @@
+  def initialize(content_item, country_name = "England")
+    super(content_item)
+    @country_name = country_name
+  end
```
👍

Rosa-Fox commented on app/presenters/local_transaction_presenter.rb:
```diff
@@ -1,13 +1,53 @@
+  def unavailable?
+    return false if devolved_administration_availability_is_not_present
+
+    send("#{key}_availability")["type"] == "unavailable"
```
👍

Rosa-Fox commented on app/presenters/local_transaction_presenter.rb:
```diff
@@ -1,13 +1,53 @@
+    send("#{key}_availability")["type"] == "unavailable"
+  end
+
+  def devolved_administration?
```
👍

Rosa-Fox commented on app/presenters/local_transaction_presenter.rb:
```diff
@@ -1,13 +1,53 @@
+    send("#{key}_availability")["type"] == "devolved_administration_service"
+  end
+
+  def devolved_administration_alternative_url
+    return if devolved_administration_availability_is_not_present
+
+    if devolved_administration?
+      send("#{key}_availability")["alternative_url"]
+    end
+  end
```
👍

Rosa-Fox commented on app/presenters/local_transaction_presenter.rb:
```diff
@@ -1,13 +1,53 @@
+
+private
+
+  def key
```
👍

Rosa-Fox commented on app/views/local_transaction/devolved_administration_service.html.erb:
```diff
@@ -0,0 +1,24 @@
+    <div class="local-authority-result"
+         data-module="auto-track-event"
+         data-track-category="userAlerts:local_transaction"
+         data-track-action="ResultsShownAvailableDa">
```
I changed the value to start with a lower case character. I did suggest some values to Harriet but she chose the ones I added so I think we should keep them? Otherwise I can try asking her what she thinks about those options?

kevindew replied:
Yeah I think we should check these suggestions first before or at least give Harriet an opportunity to refine a bit. I'll drop a quick email on the thread to see how these go.

Rosa-Fox commented on app/views/local_transaction/unavailable_service.html.erb:
```diff
@@ -8,27 +8,34 @@
       We've matched the postcode to <span class="local-authority"><%= @local_authority.name %></span>.
```
Thanks. I added to the locale file along with 'to'.

Rosa-Fox commented on app/views/local_transaction/unavailable_service.html.erb:
```diff
@@ -8,27 +8,34 @@
+        <p class="govuk-body">We don't have a link for their website. Try the <a href="/find-local-council">local council search</a> instead.</p>
+      </div>
+    <% end %>
+  </div>
```
Thanks, I moved it to a translation file.

kevindew commented on test/integration/local_transactions_test.rb:
```diff
@@ -365,8 +367,8 @@ class LocalTransactionsTest < ActionDispatch::IntegrationTest
     should "show advisory message that we have no url" do
-      assert page.has_content?("We don't have a link for their website.")
-      assert page.has_link?("local council search", href: "/find-local-council")
+      content = "We don't have a link for their website. Try the &lt;a href=\"/find-local-council\"&gt;local council search&lt;/a&gt; instead."
```
Oh you're actually asserting that plain HTML is output here rather than actual HTML. It should be asserting against actual HTML - or perhaps better to do an I18n lookup.


(if you get two very similar comments on this it's because I wrote one and submitted it and GitHub no longer seems to have it)

Rosa-Fox replied:
👍

kevindew reviewed (commented):
Looking great @Rosa-Fox just a few things I flagged left.

Rosa-Fox commented on app/views/electoral/results.html.erb:
```diff
@@ -8,7 +8,7 @@
   <% if @presenter.electoral_services.present? %>
     <p class="govuk-body">
-      <%= t('formats.local_transaction.matched_postcode') %> <strong><%= @postcode.sanitized_postcode %></strong> to <strong><%= @presenter.electoral_service_name %></strong>.
+      <%= t('formats.local_transaction.matched_postcode') %> <strong><%= @postcode.sanitized_postcode %><%= @presenter.electoral_service_name %></strong>.
```
I think this one might do due to it being electoral service related?

Rosa-Fox replied:
I updated it because it used the same translation key but it's sort of unrelated. I could split into another commit potentially?

kevindew replied:
Oh yes, sorry I hadn't appreciated that this was a different controller. Yes this should have its own entry in the locale file with the other electoral keys and include the postcode.

Rosa-Fox replied:
Thanks. I made a separate key in the `electoral` section.

Rosa-Fox commented on config/locales/en.yml:
```diff
@@ -160,8 +160,9 @@ en:
-      matched_postcode: We've matched the postcode
+      matched_postcode: We've matched the postcode to
       no_local_authority: We couldn't find a council for this postcode.
+      no_local_authority_url: We don't have a link for their website. Try the <a href="/find-local-council">local council search</a> instead.
```
ah yeah its so obvious now I see it 😅 . I have updated the views/tests to use the locales.

Rosa-Fox commented:
Thanks @kevindew. It is ready for re-re-review. I updated the screen grabs in the PR description.

kevindew commented on config/locales/en.yml:
```diff
@@ -106,7 +106,9 @@ en:
       title: Get help with electoral registration
     service:
-      description: For questions about your poll card, polling place, or about returning your postal voting ballot, contact your council."
+      matched_postcode_html: We've matched the postcode to <strong>%{postcode} %{electoral_service_name}</strong>.
```
Looking at the [existing code](https://github.com/alphagov/frontend/blob/48268f18f18c12ba3c33f77be551d1e59829bc9f/app/views/electoral/results.html.erb#L11) I think this should be:

```suggestion
      matched_postcode_html: We've matched the postcode <strong>%{postcode}</strong> to <strong>%{electoral_service_name}</strong>.
```

Rosa-Fox replied:
😱 good spot! I have changed it.

kevindew replied:
Thanks - by the way the change hasn't come through, maybe you've not pushed it or one of us has a wire crossed?

kevindew commented on test/integration/local_transactions_test.rb:
```diff
@@ -400,94 +403,141 @@ class LocalTransactionsTest < ActionDispatch::IntegrationTest
-        )
-      end
+    should "render matching the postcode to local authority" do
+      content = "We've matched the postcode to <span class=\"local-authority\">Cardiff</span>."
```
Shall we use the locale key for this?

Rosa-Fox replied:
👍 done!

kevindew commented on test/integration/local_transactions_test.rb:
```diff
@@ -400,94 +403,141 @@ class LocalTransactionsTest < ActionDispatch::IntegrationTest
+      assert page.has_content? I18n.t("formats.local_transaction.service_not_available", country_name: "Wales")
+    end
+
+    should "render matching the postcode to local authority" do
+      content = "We've matched the postcode to <span class=\"local-authority\">Cardiff</span>."
+      assert page.body.include?(content)
+    end
+
+    should "not link to the authority" do
+      assert page.has_no_link?("Go to their website")
+    end
```
Similar point about I18n to above.

Rosa-Fox replied:
👍 done!

kevindew reviewed (commented):
Looks super close now Rosa, just a few more snags. I've still got a chat going with Harriet to resolve the analytics naming. Hopefully will sort that by your return.

Rosa-Fox commented:
Thanks @kevindew it is ready for review again. I updated the commit message to explain why the `more_information` section isn't rendered. I also updated the GA values and added the rationale from your email into the commit.

kevindew commented on app/views/local_transaction/devolved_administration_service.html.erb:
```diff
@@ -0,0 +1,25 @@
+         data-module="auto-track-event"
+         data-track-category="postcodeSearch:local_transaction"
+         data-track-action="postcodeResultShown:devolvedAdministrationService"
+         data-track-label=<%= @country_name %>>
```
```suggestion
         data-track-label="<%= @country_name %>">
```

I assume you just missed this, it wouldn't matter if it wasn't for "Northern Ireland"

Rosa-Fox replied:
thanks! I thought the syntax seemed strange with the ""! Have added back in :)

kevindew commented on app/views/local_transaction/results.html.erb:
```diff
@@ -8,16 +8,16 @@
        data-track-action="postcodeResultShown"
-       data-track-label="<%= @local_authority.name %>">
-      <%= t('formats.local_transaction.matched_postcode') %> <span class="local-authority"><%= @local_authority.name %></span>.
+       data-track-label=<%= @local_authority.name %>>
```
Quotes on this too.

Rosa-Fox replied:
👍

kevindew commented on app/views/local_transaction/unavailable_service.html.erb:
```diff
@@ -1,34 +1,42 @@
+           data-module="auto-track-event"
+           data-track-category="postcodeSearch:local_transaction"
+           data-track-action="postcodeResultShown:devolvedAdministrationUnavailable"
+           data-track-label=<%= @country_name %>>
```
This event should fire regardless of whether there is a local_authority url present or not, so should be set before the conditional. (label needs quotes too)

Rosa-Fox replied:
👍

kevindew commented on app/views/local_transaction/unavailable_service.html.erb:
```diff
@@ -1,34 +1,42 @@
-    </section>
-  <% end %>
+    <% else %>
+      <div class="local-authority-result"
+      data-module="auto-track-event"
+      data-track-category="userAlerts:local_transaction"
+      data-track-action="postcodeResultShown:laMatchNoLinkNoAuthorityUrl">
+        <p class="govuk-body"><%= t('formats.local_transaction.no_local_authority_url_html') %></p>
+      </div>
```
Shall we put this in a partial so that it doesn't need to be duplicated between this and results? That seems to offer a slightly better chance that this stays up to date. Could probably call it `_no_local_authority_url.html.erb` to match the translation key.

Rosa-Fox replied:
Good idea! Have put in a partial.

kevindew reviewed (commented):
Super nearly there now! Just a few small issues to fix.

Rosa-Fox commented:
@kevindew thanks! Have made the updates. Sorry about the `matched_postcode_html` I must have messed up rewriting the commit somehow.

kevindew reviewed (approved):
All looks good to me 👍  I've given it a spin on integration and it worked well too 🎉

Rosa-Fox commented:
Thanks @kevindew :)
